n2survey/lime/survey.py

```diff
@@ -75,28 +77,113 @@ def read_responses(self, responses_file: str) -> None:
+                (self.questions["type"] == "multiple-choice")
+                & (self.questions["format"] == "longtext"),
+                "question_group",
+            ].unique()
+        )
+        for question in multiple_choice_questions:
+            other_column = question + "other"
+            question_responses.insert(
+                question_responses.columns.get_loc(other_column),
+                question + "T",
+                # Fill in new column based on "{question_id}other" column data
+                pd.Categorical(
+                    question_responses[other_column].where(
+                        question_responses[other_column].isnull(), "Y"
+                    )
+                ),
+            )
+
+        # Validate data structure
+        # Check for columns not listed in survey structure df
+        not_in_structure = list(
+            set(question_responses.columns) - set(self.questions.index)
+        )
+        if not_in_structure:
+            warnings.warn(
+                f"The following columns in the data csv file are not found in the survey structure and are dropped:\n{not_in_structure}"
+            )
+            question_responses = question_responses.drop(not_in_structure, axis=1)
+        # Ceheck for questions not listed in data csv
+        not_in_data = list(set(self.questions.index) - set(question_responses.columns))
+        if not_in_structure:
+            warnings.warn(
+                f"The following questions in the survey structure are not found in the data csv file:\n{not_in_data}"
+            )
+
+        self.responses = question_responses
+        self.sys_info = system_info
+
+    def _get_dtype_info(self, columns, renamed_columns):
+        """Get dtypes for columns in data csv
 
-        # Process the data (optional for now)
-        # * Optimize dtypes. At least for category fields
-        # the dtype can be optimized using category.
-        # * Clean values. Make sure that NA, None, etc. parsed properly
-        # TODO: ...
+        Args:
+            columns (list): List of column names from data csv
+            renamed_columns (list): List of column names modified to match self.questions entries
 
-        # Save to `self`
-        # TODO: self.responses = responses
+        Returns:
+            dict: Dictionary of column names and dtypes
+            list: List of datetime columns
+        """
 
-        raise NotImplementedError()
+        # Compile dict with dtype for each column and list of datetime columns
+        dtype_dict = {}
+        datetime_columns = []
+        for column, renamed_column in zip(columns, renamed_columns):
+            if renamed_column in self.questions.index:
+                if (
+                    self.questions.loc[renamed_column, "type"]
+                    in ["single-choice", "multiple-choice", "array"]
+                    and self.questions.loc[renamed_column, "format"] is None
+                ):
```

Comment:
Wouldn't `self.questions.loc[renamed_column, "choices"] is not None` be the same thing?

Reply:
![image](https://user-images.githubusercontent.com/77344869/139712899-428afa99-8f66-4d4d-a597-66d5b0d30b9a.png)

![image](https://user-images.githubusercontent.com/77344869/139712825-7f6054c9-0be0-4584-8088-b4fadd40b905.png)

I'm a bit surprised too that it actually is not. If I print out all questions that satisfy your condition and not mine, I get all the contingent "other" questions (see screenshot 1). Apparently "nan" given in the "choices" is not None?

Reply:
Oh.. I guess the problem is the way we check. In pandas, one should use [notnull](https://friedolin.uni-jena.de/qisserver/rds?state=verpublish&status=init&vmfile=no&publishid=193012&moduleCall=webInfo&publishConfFile=webInfo&publishSubDir=veranstaltung), like `df.loc[a, b].notnull()`. Because, by default, missing values are converted to `np.NaN` which is not the same as `None`